File: 907-koko-eating-bananas/koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    // given an array of nums and a number as inputs
    // outputs: number
    // summarize: we are given an array of nums and some time h
    // we want to find k where k is the minimum number where we can eat all piles within h hours

    // if k > piles[i] then we eat a pile => remaining pile length - 1
    // the most we can eat is max(piles) since it would mean we eat 1 pile every hour
    // if eat too slow > h => then up the speed (k)
    // if eat too fast < h => then lower the speed (k)

    // check from k = 1  to  max(piles) [1,2,3,4,5,6,7,8,9,10,11]


    const checkSpeed = (k, h, piles) => {
        let speed = 0
        for (let num of piles) {
            speed += Math.ceil(num / k)
        }
        return speed
    }

    let l = 1
    let r = Math.max(...piles)
    let minK = 0

    while (l <= r) {
        const mid = Math.floor((l + r) / 2)
        const speed = checkSpeed(mid, h, piles)
        if (speed <= h) {
            minK = mid
            r = mid - 1
        } else {
            l = mid + 1
        }
    } 

    return minK

};

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # return minimum k such that all bananas are gone within h hours

        # Determine an eating speed such that all bananas will be gone within h hours
        # [4,11,20,23,30]

        def hours_to_finish(eating_speed):
            hours = 0
            for i,p in enumerate(piles):
                if p % eating_speed == 0:
                    hours += p // eating_speed
                else:
                    hours += p // eating_speed + 1
            return hours



        max_bananas = max(piles)
        l = 1
        r = max_bananas

        while l < r:
            mid = (l+r) // 2
            if hours_to_finish(mid) > h:
                l = mid + 1
            else:
                r = mid

        return l